pkg/apis/batch/types.go

```diff
@@ -149,6 +162,28 @@ type JobSpec struct {
 	// TTLAfterFinished feature.
 	// +optional
 	TTLSecondsAfterFinished *int32
+
+	// CompletionMode specifies how Pod completions are tracked. It can be
+	// `NonIndexed` (default) or `Indexed`.
+	//
+	// `NonIndexed` means that each Pod completion is homologous to each other.
+	// The Job is considered complete when there have been .spec.completions
+	// successfully completed Pods.
+	//
+	// `Indexed` means that each Pod completion needs to be tracked individually.
+	// Each Pods gets an associated completion index, which is available in the
+	// annotation `alpha.batch.k8s.io/job-completion-index`.
+	// The Job is considered complete when there is one successfully completed Pod
+	// for each // index in the range 0 to (.spec.completions - 1).
+	// When value is `Indexed`, .spec.completions must be specified and
+	// `.spec.parallelism` must be less than or equal to 10^5.
+	//
+	// This field is alpha-level and is only honored by servers that enable the
+	// IndexedJob feature gate. More completion modes can be added in the future.
+	// If the Job controller observes a mode that it doesn't recognize, it manages
```

Comment:
why not reject it at validation?

Reply:
We do reject it at validation. But an apiserver can temporarily be at a newer version than the controller. Then the controller might see a mode that it doesn't recognize.

This note is a safeguard to allow the addition of new modes in the future. In production clusters, the situation described wouldn't happen, as we first release new modes with a disabled feture gate.

But actually, it's safer to skip the sync. Changing.